Lib/collections/__init__.py

```diff
@@ -788,7 +788,21 @@ def __eq__(self, other):
         'True if all counts agree. Missing counts are treated as zero.'
         if not isinstance(other, Counter):
             return NotImplemented
-        return all(self[e] == other[e] for c in (self, other) for e in c)
+
+        if super().__eq__(other):
```

Comment:
`__eq__()` can return `NotImplemented`, and `NotImplemented` should not be used in boolean context (it emits a warning).

If you say that `dict.__eq__()` never returns `NotImplemented` for arguments that are instances of dict subclasses, than use `dict.__eq__()`.

I wonder what effect this optimization has on comparison of two large Counters with equal size but slightly different keys?
```py
c1 = Counter(large_dict)
c2 = Counter(large_dict)
c1[new_k1] = 0
c2[new_k2] = v  # may be 0 or not 0
```

Reply:
> If you say that dict.__eq__() never returns NotImplemented for arguments that are instances of dict subclasses, than use dict.__eq__().

Because of the `return NotImplemented` for non-`Counters`, I believe we only need to be concerned with the case that `dict.__eq__` works for two `Counter` instances. I'm not able to think of a situation in which this would fail.

Reply:
Something like:
```py
class A(dict):
    def __eq__(s, o): return NotImplemented

class B(Counter, A): pass
```

Reply:
Here's a little test script. Please feel free to probe further if you have other ideas. Both main and this branch return the same results.

```python
from collections import Counter


class A(dict):
    def __eq__(s, o): return NotImplemented

class B(Counter, A):
    pass

class C(A, Counter):
    pass


a = A()
b = B()
c = C()

print(a == b)
print(b == a)
print(c == b)
print(b == c)
print(a == c)
print(c == a)

print({} == a)
print(a == {})

print({} == b)
print(b == {})

print({} == c)
print(c == {})
```

result
```
False
False
True
True
False
False
True
True
True
True
True
True

```

Reply:
```pycon
>>> print(c == b)
/home/serhiy/py/cpython/Lib/collections/__init__.py:792: DeprecationWarning: NotImplemented should not be used in a boolean context
  if super().__eq__(other):
True
>>> print(b == c)
/home/serhiy/py/cpython/Lib/collections/__init__.py:792: DeprecationWarning: NotImplemented should not be used in a boolean context
  if super().__eq__(other):
True
```

Reply:
The simple solution would be just to check `is True` rather than the implicit bool. Does that meaningfully alter the performance gains here?

Reply:
@AA-Turner No, it's essentially unchanged, so I added the explicit `is True`. Results from the benchmark [here](https://github.com/python/cpython/pull/113948#issuecomment-1898848244) after the change are:
```
0: 0.09299182891845703
1: 0.20043516159057617
2: 0.08538293838500977
3: 0.7197530269622803
4: 0.719200611114502
5: 0.017891883850097656
total duration: 1.8356554508209229
```

This is presumably because `is True` is only done once per call -- it's not part of any iteration.